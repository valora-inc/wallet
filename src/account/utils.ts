import { parsePhoneNumber } from '@celo/phone-utils'

const ADDRESS_LENGTH = 42 // TODO(ACT-1173): see if this can be replaced with a viem helper

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}

export function getPhoneNumberDetails(
  phoneNumber: string,
  countryCallingCode: string,
  countryCodeAlpha2: string
) {
  const phoneDetails = parsePhoneNumber(phoneNumber, countryCallingCode)

  if (phoneDetails) {
    return {
      // Show international display number to avoid confusion
      internationalPhoneNumber: phoneDetails.displayNumberInternational,
      e164Number: phoneDetails.e164Number,
      isValidNumber: true,
      countryCodeAlpha2: phoneDetails.regionCode!,
    }
  } else {
    return {
      internationalPhoneNumber: phoneNumber,
      e164Number: '',
      isValidNumber: false,
      countryCodeAlpha2,
    }
  }
}
