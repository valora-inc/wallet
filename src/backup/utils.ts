import {
  CELO_DERIVATION_PATH_BASE,
  generateKeys,
  MnemonicLanguages,
} from '@celo/cryptographic-utils'
import CryptoJS from 'crypto-js'
import { useAsync } from 'react-async-hook'
import * as bip39 from 'react-native-bip39'
import { showError } from 'src/alert/actions'
import { OnboardingEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { getPassword } from 'src/pincode/authentication'
import { useDispatch, useSelector } from 'src/redux/hooks'
import { removeStoredItem, retrieveStoredItem, storeItem } from 'src/storage/keychain'
import Logger from 'src/utils/Logger'
import { ETHEREUM_DERIVATION_PATH } from 'src/web3/consts'
import { currentAccountSelector } from 'src/web3/selectors'

const TAG = 'Backup/utils'

const MNEMONIC_STORAGE_KEY = 'mnemonic'

export async function generateKeysFromMnemonic(mnemonic: string) {
  const wordCount = countMnemonicWords(mnemonic)
  const derivationPath = wordCount === 24 ? CELO_DERIVATION_PATH_BASE : ETHEREUM_DERIVATION_PATH
  return generateKeys(mnemonic, undefined, undefined, undefined, bip39, derivationPath)
}

export function getMnemonicLanguage(language: string | null) {
  switch (language?.slice(0, 2)) {
    case 'es': {
      return MnemonicLanguages.spanish
    }
    case 'pt': {
      return MnemonicLanguages.portuguese
    }
    default: {
      return MnemonicLanguages.english
    }
  }
}

export async function storeMnemonic(mnemonic: string, account: string | null, password?: string) {
  if (!account) {
    throw new Error('Account not yet initialized')
  }
  const passwordToUse = password ?? (await getPassword(account))
  const encryptedMnemonic = await encryptMnemonic(mnemonic, passwordToUse)
  return storeItem({ key: MNEMONIC_STORAGE_KEY, value: encryptedMnemonic })
}

export async function clearStoredMnemonic() {
  await removeStoredItem(MNEMONIC_STORAGE_KEY)
}

export async function getStoredMnemonic(
  account: string | null,
  password?: string
): Promise<string | null> {
  try {
    if (!account) {
      throw new Error('Account not yet initialized')
    }

    Logger.debug(TAG, 'Checking keystore for mnemonic')
    const encryptedMnemonic = await retrieveStoredItem(MNEMONIC_STORAGE_KEY)
    if (!encryptedMnemonic) {
      throw new Error('No mnemonic found in storage')
    }

    const passwordToUse = password ?? (await getPassword(account))
    return decryptMnemonic(encryptedMnemonic, passwordToUse)
  } catch (error) {
    Logger.error(TAG, 'Failed to retrieve mnemonic', error)
    return null
  }
}

export function onGetMnemonicFail(viewError: (error: ErrorMessages) => void, context?: string) {
  viewError(ErrorMessages.FAILED_FETCH_MNEMONIC)
  ValoraAnalytics.track(OnboardingEvents.backup_error, {
    error: 'Failed to retrieve Recovery Phrase',
    context,
  })
}

export function useAccountKey(): string | null {
  const dispatch = useDispatch()
  const account = useSelector(currentAccountSelector)
  const asyncAccountKey = useAsync(getStoredMnemonic, [account])

  if (!asyncAccountKey || asyncAccountKey.error) {
    onGetMnemonicFail((error) => dispatch(showError(error)), 'useAccountKey')
  }

  return asyncAccountKey.result || null
}

export function countMnemonicWords(phrase: string): number {
  return [...phrase.trim().split(/\s+/)].length
}

// Because of a RN bug, we can't fully clean the text as the user types
// https://github.com/facebook/react-native/issues/11068
export function formatBackupPhraseOnEdit(phrase: string) {
  return phrase.replace(/\s+/gm, ' ')
}

export function isValidBackupPhrase(phrase: string) {
  const allowedPhraseLengths = [12, 24]
  const phraseLength = countMnemonicWords(formatBackupPhraseOnEdit(phrase))
  return !!phrase && allowedPhraseLengths.includes(phraseLength)
}

export async function encryptMnemonic(phrase: string, password: string) {
  return CryptoJS.AES.encrypt(phrase, password).toString()
}

export async function decryptMnemonic(encryptedMnemonic: string, password: string) {
  const bytes = CryptoJS.AES.decrypt(encryptedMnemonic, password)
  return bytes.toString(CryptoJS.enc.Utf8)
}
