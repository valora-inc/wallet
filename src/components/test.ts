function groupNumber(value: string) {
  return value.replace(/\B(?=(\d{3})+(?!\d))(?<!\.\d*)/g, 'group')
}

function actualTest(value: string, decimalSeparator: string, groupingSeparator: string) {
  return groupNumber(value).replaceAll('.', decimalSeparator).replaceAll('group', groupingSeparator)
}

console.log(actualTest('1.234.567,9999', ',', '.')) // 1,234,567.9999
