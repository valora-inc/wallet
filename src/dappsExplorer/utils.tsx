import { DappV2WithCategoryNames } from 'src/dapps/types'

/***
 * Calculates a score based on the search term and dapp
 * - +2 if the name matches the search term in a case sensitive way
 * - +1 if the name matches the search term in a case insensitive way
 * - +1 if the description matches the search term in a case insensitive way
 * - +1 if the category matches the search term in a case insensitive way
 * @param dapp: dapp with categoryNames
 * @param searchTerm: the user input search string
 * @returns {number} the score of the dapp 0 - 5
 */
export const calculateSearchScore = (dapp: DappV2WithCategoryNames, searchTerm: string) => {
  if (searchTerm === '') return 0
  const nameMatchingCaseScore = dapp.name.includes(searchTerm) ? 2 : 0
  const nameScore = dapp.name.toLowerCase().includes(searchTerm.toLowerCase()) ? 1 : 0
  const descriptionScore = dapp.description.toLowerCase().includes(searchTerm.toLowerCase()) ? 1 : 0
  const categoryScore = dapp.categoryNames.some((category) =>
    category.toLowerCase().includes(searchTerm.toLowerCase())
  )
    ? 1
    : 0
  return nameMatchingCaseScore + nameScore + descriptionScore + categoryScore
}
