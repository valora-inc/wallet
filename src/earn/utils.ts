import BigNumber from 'bignumber.js'
import { getFeatureGate } from 'src/statsig'
import { StatsigFeatureGates } from 'src/statsig/types'
import { SwapTransaction } from 'src/swap/types'
import { Network, NetworkId } from 'src/transactions/types'
import { INTERNAL_RPC_SUPPORTED_NETWORKS } from 'src/viem'
import { networkIdToNetwork } from 'src/web3/networkConfig'

/**
 * Helper function to check if gas is subsidized for the given networkId
 * @param networkId - networkId to check
 * @returns boolean
 */
export function isGasSubsidizedForNetwork(networkId: NetworkId) {
  return (
    (INTERNAL_RPC_SUPPORTED_NETWORKS as readonly Network[]).includes(
      networkIdToNetwork[networkId]
    ) && getFeatureGate(StatsigFeatureGates.SUBSIDIZE_STABLECOIN_EARN_GAS_FEES)
  )
}

/**
 * Gets the swap to amount in decimals given a swap transaction and the swap
 * from amount in decimals. This could also be taken from the buy amount, but
 * that is in wei and we would need the token decimals. This follows what is
 * done in the SwapScreen
 * @param object.swapTransaction - swap transaction object
 * @param object.fromAmount - swap from amount in decimals (usually from the user input)
 * @returns swap to amount in decimals
 */
export function getSwapToAmountInDecimals({
  swapTransaction,
  fromAmount,
}: {
  swapTransaction: SwapTransaction
  fromAmount: BigNumber
}) {
  return fromAmount.multipliedBy(swapTransaction.price)
}
