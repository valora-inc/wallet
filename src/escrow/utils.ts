import { EscrowedPayment } from 'src/escrow/actions'
import { identifierToE164NumberSelector } from 'src/identity/selectors'
import { ContactRecipient } from 'src/recipients/recipient'
import { useSelector } from 'src/redux/hooks'

export const useEscrowPaymentRecipient = (payment: EscrowedPayment): ContactRecipient => {
  const { recipientPhone, recipientIdentifier } = payment

  const identifierToE164Number = useSelector(identifierToE164NumberSelector)
  const recipientCache = useSelector((state) => state.recipients.phoneRecipientCache)
  const phoneNumber = identifierToE164Number[recipientIdentifier] ?? recipientPhone
  const recipient = recipientCache[phoneNumber] ?? {
    name: phoneNumber,
    e164PhoneNumber: phoneNumber,
    contactId: '',
  }

  return recipient
}
