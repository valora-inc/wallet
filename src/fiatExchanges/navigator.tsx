import { navigate, navigateHome } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { getExperimentParams } from 'src/statsig'
import { ExperimentConfigs } from 'src/statsig/constants'
import { StatsigExperiments } from 'src/statsig/types'

export const navigateToFiatExchangeStart = () => {
  if (
    getExperimentParams(ExperimentConfigs[StatsigExperiments.HOME_SCREEN_ACTIONS])
      .showAddWithdrawOnMenu
  ) {
    navigate(Screens.FiatExchange)
  } else {
    navigateHome()
  }
}
