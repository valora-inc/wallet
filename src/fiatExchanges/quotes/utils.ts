import { SettlementEstimation, SettlementTime } from 'src/fiatExchanges/quotes/constants'
import Logger from 'src/utils/Logger'

const TAG = 'fiatExchanges/quotes/utils'

export const getSettlementTimeString = (
  settlementEstimation: SettlementEstimation,
  settlementTimeStrings: Record<SettlementTime, string>
) => {
  switch (settlementEstimation.settlementTime) {
    case SettlementTime.LESS_THAN_ONE_HOUR:
      return { timeString: settlementTimeStrings[SettlementTime.LESS_THAN_ONE_HOUR] }
    case SettlementTime.LESS_THAN_X_HOURS:
    case SettlementTime.LESS_THAN_X_DAYS:
      return {
        timeString: settlementTimeStrings[settlementEstimation.settlementTime],
        upperBound: settlementEstimation.upperBound,
      }
    case SettlementTime.X_TO_Y_HOURS:
    case SettlementTime.X_TO_Y_DAYS:
      return {
        timeString: settlementTimeStrings[settlementEstimation.settlementTime],
        lowerBound: settlementEstimation.lowerBound,
        upperBound: settlementEstimation.upperBound,
      }
    default:
      // Should not be reachable
      Logger.warn(TAG, 'Unexpected settlement time value in settlement estimation')
      return {}
  }
}
