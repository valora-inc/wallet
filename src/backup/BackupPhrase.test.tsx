import { render } from '@testing-library/react-native'
import * as React from 'react'
import 'react-native'
import * as Keychain from 'react-native-keychain'
import { Provider } from 'react-redux'
import BackupPhrase from 'src/backup/BackupPhrase'
import { Screens } from 'src/navigator/Screens'
import { getExperimentParams } from 'src/statsig'
import { createMockStore, getMockStackScreenProps } from 'test/utils'

jest.mock('src/statsig', () => ({
  getExperimentParams: jest.fn().mockReturnValue({
    useNewBackupFlowCopy: false,
  }),
}))

it('renders correctly with backup not completed', () => {
  const tree = render(
    <Provider store={createMockStore()}>
      <BackupPhrase {...getMockStackScreenProps(Screens.BackupPhrase)} />
    </Provider>
  )
  expect(tree).toMatchSnapshot()
})

it('renders correctly with backup not completed and using new backup flow copy', () => {
  ;(getExperimentParams as jest.Mock).mockReturnValueOnce({
    useNewBackupFlowCopy: true,
  })
  const tree = render(
    <Provider store={createMockStore()}>
      <BackupPhrase {...getMockStackScreenProps(Screens.BackupPhrase)} />
    </Provider>
  )
  expect(tree).toMatchSnapshot()
})

it('renders correctly with backup completed', () => {
  const tree = render(
    <Provider store={createMockStore({ account: { backupCompleted: true } })}>
      <BackupPhrase {...getMockStackScreenProps(Screens.BackupPhrase)} />
    </Provider>
  )
  expect(tree).toMatchSnapshot()
})

it('renders correctly with backup completed and using new backup flow copy', () => {
  ;(getExperimentParams as jest.Mock).mockReturnValueOnce({
    useNewBackupFlowCopy: true,
  })
  const tree = render(
    <Provider store={createMockStore()}>
      <BackupPhrase {...getMockStackScreenProps(Screens.BackupPhrase)} />
    </Provider>
  )
  expect(tree).toMatchSnapshot()
})

it('still renders when mnemonic doesnt show up', () => {
  const mockGetGenericPassword = Keychain.getGenericPassword as jest.Mock
  mockGetGenericPassword.mockResolvedValue(null)

  const tree = render(
    <Provider store={createMockStore({ account: { backupCompleted: true } })}>
      <BackupPhrase {...getMockStackScreenProps(Screens.BackupPhrase)} />
    </Provider>
  )
  expect(tree).toMatchSnapshot()
})
