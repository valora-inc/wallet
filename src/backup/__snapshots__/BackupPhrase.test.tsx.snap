// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with backup completed 1`] = `
<RNCSafeAreaView
  edges={
    [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    {
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 56,
          "justifyContent": "space-between",
          "overflow": "hidden",
          "width": "100%",
        },
        {
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "height": "100%",
          "justifyContent": "center",
        }
      }
    >
      <Modal
        animationType="none"
        deviceHeight={1334}
        deviceWidth={null}
        hardwareAccelerated={false}
        hideModalContentWhileAnimating={false}
        onBackdropPress={[Function]}
        onModalHide={[Function]}
        onModalWillHide={[Function]}
        onModalWillShow={[Function]}
        onRequestClose={[Function]}
        panResponderThreshold={4}
        scrollHorizontal={false}
        scrollOffset={0}
        scrollOffsetMax={0}
        scrollTo={null}
        statusBarTranslucent={true}
        supportedOrientations={
          [
            "portrait",
            "landscape",
          ]
        }
        swipeThreshold={100}
        transparent={true}
        visible={false}
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        hitSlop={
          {
            "bottom": 10,
            "left": 10,
            "right": 10,
            "top": 10,
          }
        }
        nativeBackgroundAndroid={
          {
            "attribute": "selectableItemBackgroundBorderless",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="CancelButton"
      >
        <Text
          style={
            [
              {
                "color": "#1AB775",
                "fontFamily": "Inter-Regular",
                "fontSize": 16,
                "lineHeight": 22,
              },
              [
                {
                  "color": "#2E3338",
                },
                {
                  "color": "#666666",
                },
              ],
            ]
          }
        >
          cancel
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "height": "100%",
          "justifyContent": "center",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        hitSlop={
          {
            "bottom": 10,
            "left": 10,
            "right": 10,
            "top": 10,
          }
        }
        nativeBackgroundAndroid={
          {
            "attribute": "selectableItemBackgroundBorderless",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <Text
          style={
            {
              "color": "#1AB775",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          moreInfo
        </Text>
      </View>
    </View>
  </View>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
        "padding": 16,
      }
    }
  >
    <View>
      <View>
        <View
          style={
            {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingHorizontal": 2,
            }
          }
        >
          <View
            style={
              {
                "flexDirection": "column",
              }
            }
          >
            <Text
              style={
                {
                  "color": "#2E3338",
                  "fontFamily": "Jost-Medium",
                  "fontSize": 22,
                  "lineHeight": 28,
                  "marginBottom": 16,
                }
              }
            >
              writeDownKeyExperimental
            </Text>
          </View>
        </View>
        <View
          accessibilityLabel=""
          style={
            {
              "alignContent": "center",
              "backgroundColor": "#F1F0EB",
              "borderRadius": 8,
              "flexDirection": "row",
              "flexWrap": "wrap",
              "justifyContent": "center",
              "marginTop": 8,
              "padding": 8,
            }
          }
          testID="AccountKeyWordsContainer"
        />
      </View>
      <Text
        style={
          {
            "color": "#2E3338",
            "fontFamily": "Inter-Regular",
            "fontSize": 16,
            "lineHeight": 22,
            "marginTop": 16,
          }
        }
      >
        backupKeyWarning
      </Text>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;

exports[`renders correctly with backup not completed 1`] = `
<RNCSafeAreaView
  edges={
    [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    {
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 56,
          "justifyContent": "space-between",
          "overflow": "hidden",
          "width": "100%",
        },
        {
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "height": "100%",
          "justifyContent": "center",
        }
      }
    >
      <Modal
        animationType="none"
        deviceHeight={1334}
        deviceWidth={null}
        hardwareAccelerated={false}
        hideModalContentWhileAnimating={false}
        onBackdropPress={[Function]}
        onModalHide={[Function]}
        onModalWillHide={[Function]}
        onModalWillShow={[Function]}
        onRequestClose={[Function]}
        panResponderThreshold={4}
        scrollHorizontal={false}
        scrollOffset={0}
        scrollOffsetMax={0}
        scrollTo={null}
        statusBarTranslucent={true}
        supportedOrientations={
          [
            "portrait",
            "landscape",
          ]
        }
        swipeThreshold={100}
        transparent={true}
        visible={false}
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        hitSlop={
          {
            "bottom": 10,
            "left": 10,
            "right": 10,
            "top": 10,
          }
        }
        nativeBackgroundAndroid={
          {
            "attribute": "selectableItemBackgroundBorderless",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="CancelButton"
      >
        <Text
          style={
            [
              {
                "color": "#1AB775",
                "fontFamily": "Inter-Regular",
                "fontSize": 16,
                "lineHeight": 22,
              },
              [
                {
                  "color": "#2E3338",
                },
                {
                  "color": "#666666",
                },
              ],
            ]
          }
        >
          cancel
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "height": "100%",
          "justifyContent": "center",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        hitSlop={
          {
            "bottom": 10,
            "left": 10,
            "right": 10,
            "top": 10,
          }
        }
        nativeBackgroundAndroid={
          {
            "attribute": "selectableItemBackgroundBorderless",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <Text
          style={
            {
              "color": "#1AB775",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          moreInfo
        </Text>
      </View>
    </View>
  </View>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
        "padding": 16,
      }
    }
  >
    <View>
      <View>
        <View
          style={
            {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingHorizontal": 2,
            }
          }
        >
          <View
            style={
              {
                "flexDirection": "column",
              }
            }
          >
            <Text
              style={
                {
                  "color": "#2E3338",
                  "fontFamily": "Jost-Medium",
                  "fontSize": 22,
                  "lineHeight": 28,
                  "marginBottom": 16,
                }
              }
            >
              writeDownKeyExperimental
            </Text>
          </View>
        </View>
        <View
          accessibilityLabel=""
          style={
            {
              "alignContent": "center",
              "backgroundColor": "#F1F0EB",
              "borderRadius": 8,
              "flexDirection": "row",
              "flexWrap": "wrap",
              "justifyContent": "center",
              "marginTop": 8,
              "padding": 8,
            }
          }
          testID="AccountKeyWordsContainer"
        />
      </View>
      <Text
        style={
          {
            "color": "#2E3338",
            "fontFamily": "Inter-Regular",
            "fontSize": 16,
            "lineHeight": 22,
            "marginTop": 16,
          }
        }
      >
        backupKeyWarning
      </Text>
    </View>
  </RCTScrollView>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginVertical": 16,
        "paddingHorizontal": 16,
      }
    }
  >
    <AndroidSwitch
      accessibilityRole="switch"
      enabled={true}
      on={false}
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="backupKeySavedSwitch"
      thumbTintColor="#E6E6E6"
      trackColorForFalse="#757575"
      trackColorForTrue="#1AB775"
      trackTintColor="#757575"
    />
    <Text
      onPress={[Function]}
      style={
        {
          "color": "#2E3338",
          "flex": 1,
          "fontFamily": "Inter-Regular",
          "fontSize": 16,
          "lineHeight": 22,
          "paddingLeft": 8,
        }
      }
    >
      savedConfirmation
    </Text>
  </View>
  <View
    style={
      [
        {
          "flexDirection": "column",
        },
        {
          "paddingBottom": 16,
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 100,
          },
        ]
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": true,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={false}
        nativeBackgroundAndroid={
          {
            "attribute": "selectableItemBackground",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "backgroundColor": "#F1F0EB",
              "flexDirection": "row",
              "flexGrow": 1,
              "height": 48,
              "justifyContent": "center",
              "opacity": undefined,
              "paddingHorizontal": 24,
              "paddingVertical": 5,
            },
            undefined,
          ]
        }
        testID="backupKeyContinue"
      >
        <Text
          maxFontSizeMultiplier={1}
          style={
            {
              "color": "#666666",
              "fontFamily": "Inter-SemiBold",
              "fontSize": 16,
              "lineHeight": 22,
              "marginLeft": 0,
              "marginRight": 0,
            }
          }
        >
          continue
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;

exports[`still renders when mnemonic doesnt show up 1`] = `
<RNCSafeAreaView
  edges={
    [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    {
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 56,
          "justifyContent": "space-between",
          "overflow": "hidden",
          "width": "100%",
        },
        {
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "height": "100%",
          "justifyContent": "center",
        }
      }
    >
      <Modal
        animationType="none"
        deviceHeight={1334}
        deviceWidth={null}
        hardwareAccelerated={false}
        hideModalContentWhileAnimating={false}
        onBackdropPress={[Function]}
        onModalHide={[Function]}
        onModalWillHide={[Function]}
        onModalWillShow={[Function]}
        onRequestClose={[Function]}
        panResponderThreshold={4}
        scrollHorizontal={false}
        scrollOffset={0}
        scrollOffsetMax={0}
        scrollTo={null}
        statusBarTranslucent={true}
        supportedOrientations={
          [
            "portrait",
            "landscape",
          ]
        }
        swipeThreshold={100}
        transparent={true}
        visible={false}
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        hitSlop={
          {
            "bottom": 10,
            "left": 10,
            "right": 10,
            "top": 10,
          }
        }
        nativeBackgroundAndroid={
          {
            "attribute": "selectableItemBackgroundBorderless",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="CancelButton"
      >
        <Text
          style={
            [
              {
                "color": "#1AB775",
                "fontFamily": "Inter-Regular",
                "fontSize": 16,
                "lineHeight": 22,
              },
              [
                {
                  "color": "#2E3338",
                },
                {
                  "color": "#666666",
                },
              ],
            ]
          }
        >
          cancel
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "height": "100%",
          "justifyContent": "center",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        hitSlop={
          {
            "bottom": 10,
            "left": 10,
            "right": 10,
            "top": 10,
          }
        }
        nativeBackgroundAndroid={
          {
            "attribute": "selectableItemBackgroundBorderless",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <Text
          style={
            {
              "color": "#1AB775",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          moreInfo
        </Text>
      </View>
    </View>
  </View>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
        "padding": 16,
      }
    }
  >
    <View>
      <View>
        <View
          style={
            {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingHorizontal": 2,
            }
          }
        >
          <View
            style={
              {
                "flexDirection": "column",
              }
            }
          >
            <Text
              style={
                {
                  "color": "#2E3338",
                  "fontFamily": "Jost-Medium",
                  "fontSize": 22,
                  "lineHeight": 28,
                  "marginBottom": 16,
                }
              }
            >
              writeDownKeyExperimental
            </Text>
          </View>
        </View>
        <View
          accessibilityLabel=""
          style={
            {
              "alignContent": "center",
              "backgroundColor": "#F1F0EB",
              "borderRadius": 8,
              "flexDirection": "row",
              "flexWrap": "wrap",
              "justifyContent": "center",
              "marginTop": 8,
              "padding": 8,
            }
          }
          testID="AccountKeyWordsContainer"
        />
      </View>
      <Text
        style={
          {
            "color": "#2E3338",
            "fontFamily": "Inter-Regular",
            "fontSize": 16,
            "lineHeight": 22,
            "marginTop": 16,
          }
        }
      >
        backupKeyWarning
      </Text>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;
